### PR TITLE
Followup fixes for FuseFunctionsPass

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

https://github.com/ClickHouse/ClickHouse/pull/42689#issuecomment-1310457729

### CODE REVIEW

**Comment by vdimir** [comment]
Not related to this PR, but found during writing tests:

```sql
SET allow_experimental_analyzer = 1;

CREATE TABLE tbl (`a` Int32, `b` Int32) ENGINE = Log;

-- Ok:
SELECT avg(b), sum(b), a AS b FROM tbl GROUP BY a;
SELECT avg(a - 2), sum(a - 2) FROM tbl GROUP BY a;

-- Not found column avg(default.tbl.a_0) in block:
SELECT avg(b), sum(b), a - 2 AS b FROM tbl GROUP BY a;

-- Not found column minus(default.tbl.a_0, 2_UInt8) in block:
SELECT avg(a - 2), sum(a - 2), a - 2 AS b FROM tbl GROUP BY a;
```

**Comment by vdimir** [comment]
Fuzzer failures related another code https://github.com/ClickHouse/ClickHouse/pull/43245

File: src/Analyzer/Passes/FuseFunctionsPass.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ class FuseFunctionsVisitor : public InDepthQueryTreeVisitor<FuseFunctionsVisitor
             /// Do not apply for `count()` with without arguments or `count(*)`, only `count(x)` is supported.
             return;
 
-        mapping[QueryTreeNodeWithHash(argument_nodes[0])].push_back(&node);
+        mapping[QueryTreeNodeWithHash(argument_nodes[0])].insert(&node);
```
Comment (novikd) [inline src/Analyzer/Passes/FuseFunctionsPass.cpp:51]: Why do we even get any duplicates here? I can't understand

**Comment by vdimir** [comment]
> AST fuzzer (ubsan) — ../src/AggregateFunctions/AggregateFunctionAvg.h:223:26: runtime error: downcast of address 0x7fc09a2e02e0 which does not point to an object [Details](https://s3.amazonaws.com/clickhouse-test-reports/43217/6241ed65e563831c10052c4e24288fd9f574d789/fuzzer_astfuzzerubsan//report.html)

Looks related, will investigate

File: src/Analyzer/Passes/FuseFunctionsPass.cpp (hunk 1 above)
Comment (vdimir) [reply src/Analyzer/Passes/FuseFunctionsPass.cpp]: Some subtrees can refer to the same node (at least after some optimizations, `AggregateFunctionsArithmericOperations` in the example below replaces both column with the same `avg` function).

```sql
EXPLAIN QUERY TREE run_passes = 1, dump_passes = 1, passes = 4
SELECT avg(b), x - 2 AS b FROM (SELECT number as x FROM numbers(1)) GROUP BY x;


Pass 1 QueryAnalysis - Resolve type for each query expression. Replace identifiers, matchers with query expressions. Perform constant folding. Evaluate scalar subqueries.
Pass 2 NormalizeCountVariants - Optimize count(literal), sum(1) into count().
Pass 3 CustomizeFunctions - Customize implementation of aggregate functions, and in functions.
Pass 4 AggregateFunctionsArithmericOperations - Extract arithmeric operations from aggregate functions.
QUERY id: 0
  PROJECTION COLUMNS
    avg(b) Float64
    b Int64
  PROJECTION
    LIST id: 1, nodes: 2
      FUNCTION id: 2, function_name: minus, function_type: ordinary, result_type: Int64
        ARGUMENTS
          LIST id: 3, nodes: 2
            FUNCTION id: 4, function_name: avg, function_type: aggregate, result_type: Float64
              ARGUMENTS
                LIST id: 5, nodes: 1
                  COLUMN id: 6, column_name: x, result_type: UInt64, source_id: 7
            CONSTANT id: 8, constant_value: UInt64_2, constant_value_type: UInt8
      FUNCTION id: 2, function_name: minus, function_type: ordinary, result_type: Int64
        ARGUMENTS
          LIST id: 3, nodes: 2
            FUNCTION id: 4, function_name: avg, function_type: aggregate, result_type: Float64
              ARGUMENTS
                LIST id: 5, nodes: 1
                  COLUMN id: 6, column_name: x, result_type: UInt64, source_id: 7
            CONSTANT id: 8, constant_value: UInt64_2, constant_value_type: UInt8
  JOIN TREE
    QUERY id: 7, is_subquery: 1
      PROJECTION COLUMNS
        x UInt64
      PROJECTION
        LIST id: 9, nodes: 1
          COLUMN id: 10, column_name: number, result_type: UInt64, source_id: 11
      JOIN TREE
        TABLE_FUNCTION id: 11, table_function_name: numbers
          ARGUMENTS
            LIST id: 12, nodes: 1
              CONSTANT id: 13, constant_value: UInt64_1, constant_value_type: UInt8
  GROUP BY
    LIST id: 14, nodes: 1
      COLUMN id: 6, column_name: x, result_type: UInt64, source_id: 7
```

**Comment by vdimir** [comment]
CI faliures:

> Stress test (asan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/43217/2b138388a97ba7580cc328856e54ab6546f3dcce/stress_test__asan_.html)

Known issue: https://github.com/ClickHouse/ClickHouse/issues/41369


> Stateless tests (tsan) [4/5] — fail: 1, passed: 915, skipped: 10 [Details](https://s3.amazonaws.com/clickhouse-test-reports/43217/2b138388a97ba7580cc328856e54ab6546f3dcce/stateless_tests__tsan__[4/5].html)


Looks like flaky test, created https://github.com/ClickHouse/ClickHouse/issues/43659

File: src/Analyzer/Passes/FuseFunctionsPass.cpp [hunk 2]
```diff
@@ -79,6 +79,16 @@ FunctionNodePtr createResolvedAggregateFunction(const String & name, const Query
     function_node->resolveAsAggregateFunction(aggregate_function, aggregate_function->getReturnType());
     function_node->getArguments().getNodes() = { argument };
 
+    function_node->getArgumentsNode() = std::make_shared<ListNode>(QueryTreeNodes{argument});
```
Comment (novikd) [inline src/Analyzer/Passes/FuseFunctionsPass.cpp:82]:
```suggestion
    function_node->getArguments().getNodes() = { argument };
```

**Review by novikd — APPROVED** [review]

File: src/Analyzer/Passes/FuseFunctionsPass.cpp [hunk 3]
```diff
@@ -79,6 +79,16 @@ FunctionNodePtr createResolvedAggregateFunction(const String & name, const Query
...
+
+    if (!parameters.empty())
+    {
+        QueryTreeNodes parameter_nodes;
+        for (const auto & param : parameters)
+            parameter_nodes.emplace_back(std::make_shared<ConstantNode>(param));
+        function_node->getParametersNode() = std::make_shared<ListNode>(std::move(parameter_nodes));
```
Comment (novikd) [inline src/Analyzer/Passes/FuseFunctionsPass.cpp:89]:
```suggestion
        function_node->getParameters().getNodes() = std::move(parameter_nodes);
```

**Comment by vdimir** [comment]
CI failures:

https://github.com/ClickHouse/ClickHouse/issues/43238
https://github.com/ClickHouse/ClickHouse/issues/41776
https://github.com/ClickHouse/ClickHouse/issues/41218
https://github.com/ClickHouse/ClickHouse/issues/43659

**Pull request merged**